### PR TITLE
Rework password generation

### PR DESCRIPTION
**Pull request by eerhardt**
- base it off InputAnnotation
- add extra generation parameters to support various password and user name generation needs

Contributes to #2633

cc @ellismg @vhvb1989 
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2717)

### CODE REVIEW

File: src/Aspire.Hosting/Utils/PasswordGenerator.cs [hunk 1]
```diff
@@ -14,55 +15,94 @@ internal static class PasswordGenerator
...
+    /// Creates a cryptographically random string.
     /// </summary>
-    /// <param name="lowerCase">The number of lowercase chars in the generated password or 0 to ignore this component.</param>
-    /// <param name="upperCase">The number of uppercase chars in the generated password or 0 to ignore this component.</param>
-    /// <param name="digit">The number of digits in the generated password or 0 to ignore this component.</param>
-    /// <param name="special">The number of special chars in the generated password or 0 to ignore this component.</param>
-    /// <returns>A cryptographically random password.</returns>
-    /// <exception cref="ArgumentOutOfRangeException">If lowerCase, upperCase, digit or special is negative or their sum is zero.</exception>
-    public static string GeneratePassword(int lowerCase, int upperCase, int digit, int special)
+    public static string Generate(int minLength,
```
Comment (eerhardt) [inline src/Aspire.Hosting/Utils/PasswordGenerator.cs:24]: @danmoseley I believe this addresses your feedback from https://github.com/dotnet/aspire/pull/2627#discussion_r1511951003. PTAL.

**Review by sebastienros — APPROVED** [review]

**Comment by davidfowl** [comment]
This looks really good. Do we need azd changes to merge this?

File: src/Aspire.Hosting/Utils/PasswordGenerator.cs [hunk 2]
```diff
@@ -14,55 +15,94 @@ internal static class PasswordGenerator
...
+        {
+            RandomNumberGenerator.GetItems(choices, destination.Slice(0, minValues));
+            destination = destination.Slice(minValues);
+        }
     }
+
+    private static string GetChoices(bool lower, bool upper, bool numeric, bool special) =>
+        (lower, upper, numeric, special) switch
+        {
+            (true, true, true, true) => LowerCaseChars + UpperCaseChars + NumericChars + SpecialChars,
```
Comment (danmoseley) [inline src/Aspire.Hosting/Utils/PasswordGenerator.cs:88]: We don't need a table here, just 4 ifs ?

File: src/Aspire.Hosting/Utils/PasswordGenerator.cs (hunk 1 above)
Comment (danmoseley) [reply src/Aspire.Hosting/Utils/PasswordGenerator.cs:24]: Yes

**Comment by eerhardt** [comment]
> Do we need azd changes to merge this?

From my testing, no, we don't need to wait.`azd` just ignores the extra properties. It respects the `minLength` though, so it will give a 22 character word of letters and numbers.

**Review by vhvb1989 — APPROVED** [review]

**Review by davidfowl — APPROVED** [review]

File: src/Aspire.Hosting/Utils/PasswordGenerator.cs (hunk 2 above)
Comment (danmoseley) [reply src/Aspire.Hosting/Utils/PasswordGenerator.cs:88]: Is lower allocations always strictly worth adding more constants? How often do we generate passwords? 
(Ofc not suggesting it actually matters what approach you use here)

Comment (eerhardt) [reply src/Aspire.Hosting/Utils/PasswordGenerator.cs:88]: There comes a point where you can't "unsee" these allocations. Following the "4 ifs" approach would look something like this:

```C#
string result = null;
if (lower)
{
    result += LowerCaseChars;
}
if (upper)
{
    result += UpperCaseChars;
}
if (numeric)
{
    result += NumericChars;
}
if (special)
{
    result += SpecialChars;
}

if (result is null)
{
    throw new ArgumentException("At least one character type must be enabled.");
}

return result;
```

By default all bools are "true", so we'd allocate 3 throw away strings by default every time. I'm pretty sure someone would see this code and submit a PR to optimize it later.

**Pull request merged**